internals/daemon/api_notices.go

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"regexp"
+
+	"github.com/canonical/x-go/strutil"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+)
+
+// Ensure custom keys are in the form "domain.com/key" (but somewhat more restrictive).
+var customKeyRegexp = regexp.MustCompile(
+	`^[a-z0-9]+(-[a-z0-9]+)*(\.[a-z0-9]+(-[a-z0-9]+)*)+(/[a-z0-9]+(-[a-z0-9]+)*)+$`)
+
+const (
+	maxNoticeKeyLength = 256
+	maxNoticeDataSize  = 4 * 1024
+)
+
+type addedNotice struct {
+	ID string `json:"id"`
+}
+
+func v1GetNotices(c *Command, r *http.Request, _ *UserState) Response {
+	query := r.URL.Query()
+
+	typeStrs := strutil.MultiCommaSeparatedList(query["types"])
+	types := make([]state.NoticeType, 0, len(typeStrs))
+	for _, typeStr := range typeStrs {
+		noticeType := state.NoticeType(typeStr)
+		if !noticeType.Valid() {
+			// Ignore invalid notice types (so requests from newer clients
+			// with unknown types succeed).
+			continue
```

Comment:
I assume this has been considered, but just checking: from my understanding ignoring future supported notice types will also ignore invalid notice strings from the CLI. I did a quick check in the client proposal to confirm we do not perform a validation step on user input.

It feels to me that if I type something that is wrong, it would be better if the system tells me so, rather than not receiving anything for that invalid notice type.

Having said this, I have seen the same behaviour with logs. If I type the wrong service string, I just do not get any logs, no error.

Any thoughts ?

Reply:
I had this originally rejected unknown types, but Gustavo said this in a [comment above](https://github.com/canonical/pebble/pull/296#discussion_r1350471123):

> This will make interacting with APIs in a backwards compatible manner harder, because someone coming from the future cannot do a selection on types that this system doesn't yet know about. It seems better to ignore such invalid types, after all we're respecting the contract and returning only the types matching the given filter. If a type is invalid, we'll have none of these.

I tried to capture that in my comment (probably not quite clear enough though).

And I think *not* rejecting them in the CLI is probably right, because then you can use an older client/CLI to talk to a newer server that supports additional notice types without it breaking.

Reply:
Yes, the argument does make sense. I guess its just a trade-off, and in general what I take from this is that enabling people that knows what they are doing in the most flexible is better than building in limitations to help people that is using the tool wrong. The latter case can be documented, while not having the former case has no workaround in the field. I think I like this, and I've learned an important consideration here.